pkg/disttask/framework/dispatcher/dispatcher.go

```diff
@@ -397,37 +401,116 @@ func (d *BaseDispatcher) replaceDeadNodesIfAny() error {
+		d.TaskNodes = append(d.TaskNodes, disttaskutil.GenerateExecID(serverInfo.IP, serverInfo.Port))
+	}
+}
+
+// BalanceSubtasks make count of subtasks on each liveNodes balanced and clean up subtasks on dead nodes.
+// TODO(ywqzzy): refine to make it easier for testing.
+func (d *BaseDispatcher) BalanceSubtasks() error {
+	// 1. find out nodes need to clean subtasks.
+	deadNodes := make([]string, 0)
+	deadNodesMap := make(map[string]bool, 0)
+	for _, node := range d.TaskNodes {
+		if !disttaskutil.MatchServerInfo(d.LiveNodes, node) {
+			deadNodes = append(deadNodes, node)
+			deadNodesMap[node] = true
+		}
+	}
+	// 2. get subtasks for each node before rebalance.
+	subtasks, err := d.taskMgr.GetSubtasksByStepAndState(d.ctx, d.Task.ID, d.Task.Step, proto.TaskStatePending)
+	if err != nil {
+		return err
+	}
+	if len(deadNodes) != 0 {
+		/// get subtask from deadNodes, since there might be some running subtasks on deadNodes.
+		/// In this case, all subtasks on deadNodes are in running/pending state.
+		subtasksOnDeadNodes, err := d.taskMgr.GetSubtasksByExecIdsAndStepAndState(d.ctx, deadNodes, d.Task.ID, d.Task.Step, proto.TaskStateRunning)
+		if err != nil {
+			return err
+		}
+		subtasks = append(subtasks, subtasksOnDeadNodes...)
+	}
+	// 3. group subtasks for each scheduler.
+	subtasksOnScheduler := make(map[string][]*proto.Subtask, len(d.LiveNodes)+len(deadNodes))
+	for _, node := range d.LiveNodes {
+		execID := disttaskutil.GenerateExecID(node.IP, node.Port)
+		subtasksOnScheduler[execID] = make([]*proto.Subtask, 0)
+	}
+	for _, subtask := range subtasks {
+		subtasksOnScheduler[subtask.SchedulerID] = append(
+			subtasksOnScheduler[subtask.SchedulerID],
+			subtask)
+	}
+	// 4. prepare subtasks that need to rebalance to other nodes.
+	averageSubtaskCnt := len(subtasks) / len(d.LiveNodes)
+	rebalanceSubtasks := make([]*proto.Subtask, 0)
+	for k, v := range subtasksOnScheduler {
+		if ok := deadNodesMap[k]; ok {
+			rebalanceSubtasks = append(rebalanceSubtasks, v...)
+			continue
+		}
+		// When no tidb scale-in/out and averageSubtaskCnt*len(d.LiveNodes) < len(subtasks),
+		// no need to send subtask to other nodes.
+		// eg: tidb1 with 3 subtasks, tidb2 with 2 subtasks, subtasks are balanced now.
+		if averageSubtaskCnt*len(d.LiveNodes) < len(subtasks) && len(d.TaskNodes) == len(d.LiveNodes) {
+			if len(v) > averageSubtaskCnt+1 {
+				rebalanceSubtasks = append(rebalanceSubtasks, v[0:len(v)-averageSubtaskCnt]...)
 			}
-			// replace local cache.
-			for k, v := range replaceNodes {
-				for m, n := range d.taskNodes {
-					if n == k {
-						d.taskNodes[m] = v
-						break
-					}
+			continue
+		}
+		if len(v) > averageSubtaskCnt {
+			rebalanceSubtasks = append(rebalanceSubtasks, v[0:len(v)-averageSubtaskCnt]...)
+		}
+	}
+	// 5. skip rebalance.
+	if len(rebalanceSubtasks) == 0 {
+		return nil
+	}
+	// 6.rebalance subtasks to other nodes.
+	rebalanceIdx := 0
+	for k, v := range subtasksOnScheduler {
```

Comment:
maybe merge this loop with below one, by first generating expected subtask count on live nodes, say `3 3 2`
then, assign subtasks to meet the expectation

ok for now.